### DISCUSSION TITLE
Carbon Web Components v2 Beta

### DISCUSSION THREAD

**Discussion by kennylam** [carbon-design-system/carbon-for-ibm-dotcom#10280] [Show and tell]
Hi all! We're excited to announce the beta release of `@carbon/web-components` v2. This major version upgrade includes - among other things - Carbon v11, Lit2, and an ongoing effort to achieve full parity with [`@carbon/react`](https://www.npmjs.com/package/@carbon/react) components. Any issues or questions anyone has can be discussed here. We'd love to get some feedback!

As this is an ongoing effort, there are still many `@carbon/react` components that haven't been migrated from v10 to v11. Please refer to the table below for component statuses.

Note - experimental features and components will not be included until they reach production status in `@carbon/react`.

Thanks!

| Component | Status |
|:-------------|:---------:|
|[Accordion](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-accordion--default) |:white_check_mark:|
|[Breadcrumb](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-breadcrumb--default)|:white_check_mark:|
|[Button](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-button--default)|:white_check_mark:|
|[Checkbox](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-checkbox--default)|:white_check_mark:|
|[CodeSnippet](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-code-snippet--inline)|:white_check_mark:|
|[ComboBox](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-combo-box--playground)|:white_check_mark:|
|[ContentSwitcher](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-content-switcher--default)|:white_check_mark:|
|[CopyButton](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-copy-button--default)|:white_check_mark:|
|[DataTable](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-datatable-basic--default)|:white_check_mark:|
|[DatePicker](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-date-picker--range-with-calendar)|:white_check_mark:|
|[Dropdown](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-dropdown--playground)|:white_check_mark:|
|[FileUploader](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-file-uploader--default)|:white_check_mark:|
|[FormGroup](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-form-group--default)|:white_check_mark:|
|[IconButton](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-icon-button--default)|:white_check_mark:|
|[InlineLoading](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-inline-loading--default)|:white_check_mark:|
|[Layer](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-layer--default)|:white_check_mark:|
|[Link](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-link--default)|:white_check_mark:|
|[Loading](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-loading--default)|:white_check_mark:|
|[Modal](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-modal--default)|:white_check_mark:|
|[MultiSelect](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-multi-select--default)|:white_check_mark:|
|[Notifications](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-notifications--inline)|:white_check_mark:|
|[NumberInput](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-number-input--default)|:white_check_mark:|
|[OrderedList](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-ordered-list--default)|:white_check_mark:|
|[OverflowMenu](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-overflow-menu--default)|:white_check_mark:|
|[Pagination](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-pagination--default)|:white_check_mark:|
[Popover](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-popover--tab-tip)|:white_check_mark:|
|[ProgressBar](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-progress-bar--playground)|:white_check_mark:|
|[ProgressIndicator](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-progress-indicator--default)|:white_check_mark:|
|[RadioButton](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-radio-button--default)|:white_check_mark:|
|[Search](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-search--default)|:white_check_mark:|
|[Select](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-select--default)|:white_check_mark:|
|[Skeleton](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-skeleton-skeleton-placeholder--default)|:white_check_mark:|
|[Slider](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-slider--default)|:white_check_mark:|
|[StructuredList](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-structured-list--default)|:white_check_mark:|
|[Tabs](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-tabs--default)|:white_check_mark:|
|[Tag](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-tag--default)|:white_check_mark:|
|[TextArea](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-text-area--default)|:white_check_mark:|
|[TextInput](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-text-input--default)|:white_check_mark:|
|[Tile](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-tile--default)|:white_check_mark:|
|[Toggle](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-toggle--default)|:white_check_mark:|
|[ToggleTip](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-toggletip--default)|:white_check_mark:|
|[Tooltip](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-tooltip--default)|:white_check_mark:|
|[UI Shell](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-ui-shell--fixed-side-nav)|:white_check_mark:|
|[UnorderedList](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/beta/carbon-web-components-v2/?path=/story/components-unordered-list--default)|:white_check_mark:|

**Comment by abdonrd** [comment]
Does this mean that `@carbon/web-components@2` will not be published until all components are in sync with `@carbon/react`?

**Reply by kennylam** [reply]
That's correct, the beta will be ongoing but no production (`latest`) until all components are in sync.

**Reply by jeffchew** [reply]
Just to clarify, the team is focused on what's currently in Carbon Web Components today and syncing that with Carbon React. Any net-new components (e.g. TreeView) will probably be a v2.x release.

**Reply by abdonrd** [reply]
Got it! Thanks both!

**Comment by abdonrd** [comment]
After update a project from `1.25.0` to `2.0.0-beta.0` with:

```
import '@carbon/web-components/es/components/ui-shell/index.js';
```

<img width="1067" alt="Screenshot 2023-03-29 at 7 48 30 PM" src="https://user-images.githubusercontent.com/1007051/228692933-02918b84-ee46-4d23-8abd-78fb963c9787.png">

**Reply by abdonrd** [reply]
Fixed if I remove the ui-shell import:

```
import '@carbon/web-components/es/components/ui-shell/index.js';
```

**Reply by abdonrd** [reply]
I'm not sure if this will solve the problem, but adding the extensions seems good to me anyway: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/10288

**Reply by kennylam** [reply]
Hm, interesting, I don't think we've encountered that one yet. But thanks for reporting and the PR!

**Reply by abdonrd** [reply]
If the context helps, it's a project in which we use Vite: https://vitejs.dev

**Reply by kennylam** [reply]
I've encountered that [issue](https://github.com/vitejs/vite/issues/279#issuecomment-635646269) in Vite as well.

**Reply by abdonrd** [reply]
Looks like an old one (May 28, 2020) and already fixed.
We are using the [Vite v4.2.1](https://github.com/vitejs/vite/releases/tag/v4.2.1) released the last week.

**Reply by abdonrd** [reply]
@kennylam if we can have a new beta that fixes the problem, we'd like to try again on our project! 😄

**Reply by abdonrd** [reply]
@kennylam @annawen1 sharing an example where `@carbon/web-components@2.0.0-beta.1` still have the same issue:

https://stackblitz.com/edit/github-gkn31f?file=app.vue

**Reply by kennylam** [reply]
@abdonrd Fixed! https://stackblitz.com/edit/vitejs-vite-5xlqv6?file=index.html

**Reply by abdonrd** [reply]
Thanks @kennylam! 👏 

But still not working with Nuxt (Vite under the hood):
- https://stackblitz.com/edit/github-gkn31f?file=app.vue

**Reply by kennylam** [reply]
Looks like the import path is wrong?

```js
import "./node_modules/@carbon/web-components/es/components/button/button.js";
```

This works.

**Reply by abdonrd** [reply]
Oh, wow!

@kennylam then probably there is an error with the `package.json` exports?

https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/3d9f4ec26821b59f609d01dc89f18663c136e8e0/packages/carbon-web-components/package.json#L13-L32

It's fixed if I manually add this to the `node_modules/@carbon/web-components/package.json`:

```diff
+     "./es/components/button/button.js": {
+       "default": "./es/components/button/button.js"
+     },
```

**Reply by abdonrd** [reply]
It also works if I just change this:

```diff
-     "./es/components/": {
-       "node": "./lib/components/",
-       "default": "./es/components/"
+     "./es/components/*": {
+       "default": "./es/components/*"
    },
```

**Reply by abdonrd** [reply]
Some examples:

- https://github.com/lit/lit/blob/main/packages/lit/package.json#L19
- https://github.com/shoelace-style/shoelace/blob/next/package.json#L12
- https://github.com/ing-bank/lion/blob/master/packages/ui/package.json#L14
- https://github.com/adobe/spectrum-web-components/blob/main/packages/button/package.json#L22

**Reply by abdonrd** [reply]
If you can solve the problem of `package.json` exports, plus these missed extensions (https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/10402), it seems to start working in our project (at least it doesn't break initially)! 🎉

**Reply by kennylam** [reply]
Yes, we're still in the process of syncing all the components with `@carbon/react`. There is a lot of housekeeping to do - such as exports, etc. These are a little lower on our list of priorities, but we'll get to them once we're done with the sync!

**Reply by abdonrd** [reply]
Is there any possibility to prioritize the exports?
So we could start testing the beta in our applications. Thanks!

**Reply by abdonrd** [reply]
I just want to mention that this was fixed in `2.0.0-beta.5`.
